source/dlangbot/app.d

```diff
@@ -277,18 +277,94 @@ void handlePR(string action, PullRequest* _pr)
+            // may include previously-merged commits with mentions to bugs long fixed.
+            auto statuses = descs.filter!(d => d.id == r.id);
+            auto status = statuses.empty ? null : statuses.front.status;
+            if (!status.among("NEW", "ASSIGNED", "REOPENED"))
+                continue;
+
+            // Only mention a PR at most once in an issue's comments.
+            // This check also allows us to safely include new issue
+            // references when new commits are pushed.
+            auto ourComments = oldComments[r.id]
+                // Look at only our comments
+                .filter!(comment => comment["creator"].get!string == bugzillaLogin)
+                // Get comment body
+                .map!(comment => comment["text"].get!string ~ "\n")
+                // Concatenate all of our comments' bodies
+                .join;
+            // Check for PR mention (see format string below)
+            if (ourComments.canFind("\n" ~ pr.htmlURL ~ "\n"))
+                continue;
+
+            auto issueComment = "@%s %s %s pull request #%d \"%s\" %s this issue:\n\n%-(%s\n\n%)\n\n%s".format(
+                pr.user.login, action == "opened" ? "created" : "updated", pr.baseRepoSlug,
+                pr.number, pr.title, r.fixed ? "fixing" : "mentioning",
+                r.commits.map!(c => "- " ~ c["commit"]["message"].get!string.replace("\n", "\n  ")),
+                pr.htmlURL,
+            );
+
+            updateBugs([r.id], issueComment, false, r.fixed ? ["pull"] : null);
+        }
+    }
+
+    // When a PR is merged, update Bugzilla issues
+    // (leave a comment with a link to the PR, and close them appropriately).
     if (action == "merged")
     {
         import std.algorithm.iteration : filter, map;
-        import std.array : array;
+        import std.algorithm.searching : canFind, all;
+        import std.array : array, assocArray, join, replace;
+        import std.regex : regex, matchAll;
 
-        auto ids = refs.filter!(r => r.fixed).map!(r => r.id).array;
-        if (ids.length)
-            closeIssues(ids,
-                "%s pull request #%d \"%s\" was merged:\n\n%s".format(
-                    pr.baseRepoSlug, pr.number, pr.title,
-                    pr.htmlURL,
-                ));
+        auto oldComments = getBugComments(refs.map!(r => r.id).array);
+
+        foreach (r; refs)
+        {
+            auto previouslyMentionedCommits = oldComments[r.id]
+                // Look at only our comments
+                .filter!(comment => comment["creator"].get!string == bugzillaLogin)
+                // Get comment body
+                .map!(comment => comment["text"].get!string)
+                // Concatenate all of our comments' bodies
+                .join
+                // Extract mentioned commit hashes (see format string below)
+                .matchAll(regex(`\n- ([0-9a-f]{40}) by `))
```

Comment:
Maybe store the regex as a static variable so that it's not constantly regenerated?

Reply:
We need this in Phobos:

https://github.com/CyberShadow/ae/blob/07a832d5c674a396372e348d6fa035a81b471f71/utils/regex.d#L28-L40

What did you have in mind? As the above (lazy-initialize it at runtime), or `static r = regex(...)` (construct the Regex object at compile-time)?

Reply:
Yeah just `static`.  `ctRegex` is too slow and untested. I wouldn't pay the cost of lazy initialization as there's no big downside to `static` in this case, is there?

> We need this in Phobos:

Yeah, but then again we should also deprecate `ctRegex`. ;-)

Reply:
> as there's no big downside to `static` in this case, is there?

The Regex construction runs at compile-time. This can still have a big impact on compilation time, though I didn't measure it.

> I wouldn't pay the cost of lazy initialization

The one branch will likely have an unmeasurable impact compared to the cost of actually matching text on the regex.

Reply:
> Yeah just `static`.

Done